File: py/transfer_entropy.py
"""Module transfer_entropy
   Provides means to compute and plot transfer entropy
"""

import numpy as np
import binning
import matplotlib.pyplot as plt
import time


def panzeri_treves_bias(prob_map, denom, k, l):
    """Computes the PT bias correction
    
    Arguments:

    Returns:
        float -- PT bias correction
    """
    freq_map = { comb: prob_map[comb] * denom for comb in prob_map.keys() if prob_map[comb] != 0}
    freq_map = binning.eliminate_bin(freq_map, range(k, k+1))
    freq_map_Y = binning.eliminate_bin(freq_map, range(k, k+l))
    N = sum(freq_map.values())
    Nb_sum1 = sum([f[0] for f in freq_map_Y.keys()])
    
    Nb_sum1_1 = 0
    for f in freq_map_Y.keys():
        Nb_sum1_1 += f[0]
    
    Nb_sum2 = sum([f[0] for f in freq_map.keys()])

    return (-1 / (2 * N * np.log(2))) * (Nb_sum1 - Nb_sum2)
            


def T1(pY_t_tmk, pY_tm1_tmk, pY_tm1_tmk_X_tm1_tml, pY_t_tmk_X_tm1_tml, k):
    """Computes transfer entropy from X to Y based on given probabilities
    
    Arguments:
        pY_t_tmk {dict of tuple: float} -- P(Y_t, Y_(t-1), Y_(t-2), ..., Y_(t-k))
        pY_tm1_tmk {dict of tuple: float} -- P(Y_(t-1), Y_(t-2), ..., Y_(t-(k+1)))
        pY_tm1_tmk_X_tm1_tml {dict of tuple: float} -- P(Y_(t-1), Y_(t-2), ..., Y_(t-k), X_(t-1), ... X_(t-l))
        pY_t_tmk_X_tm1_tml {dict of tuple: float} --  P(Y_t, Y_(t-1), ..., Y_(t-k), X_(t-1), ... X_(t-l))
    
    Returns:
        float -- transfer entropy from X to Y
    """
    H1 = 0
    for (Yt_list, val) in pY_t_tmk.items(): #p_list: Yt, Yt-1, Yt-2, ... Yt-k

        if val == 0 or pY_tm1_tmk[Yt_list[1:]] == 0:
            continue

        pcond = val / pY_tm1_tmk[Yt_list[1:]]
        if pcond > 0:
            H1 += val * np.log(pcond)/np.log(2)
    H1 = -H1
    H2 = 0
    for (YtXt_list, val) in pY_t_tmk_X_tm1_tml.items():

        if val == 0 or pY_tm1_tmk_X_tm1_tml[YtXt_list[1:]] == 0:
            continue

        pcond = val / pY_tm1_tmk_X_tm1_tml[YtXt_list[1:]]

        if pcond > 0:
            H2 += val * ((np.log(pcond)/np.log(2)) if pcond else 0)
    H2 = -H2

    res = H1 - H2
    return res if res > 0 else 0 

   
def T2(pY_t_tmk, pY_tm1_tmk, pY_tm1_tmk_X_tm1_tml, pY_t_tmk_X_tm1_tml, k):
    """Computes transfer entropy from X to Y based on given probabilities
    
    Arguments:
        pY_t_tmk {dict of tuple: float} -- P(Y_t, Y_(t-1), Y_(t-2), ..., Y_(t-k))
        pY_tm1_tmk {dict of tuple: float} -- P(Y_(t-1), Y_(t-2), ..., Y_(t-(k+1)))
        pY_tm1_tmk_X_tm1_tml {dict of tuple: float} -- P(Y_(t-1), Y_(t-2), ..., Y_(t-k), X_(t-1), ... X_(t-l))
        pY_t_tmk_X_tm1_tml {dict of tuple: float} --  P(Y_t, Y_(t-1), ..., Y_(t-k), X_(t-1), ... X_(t-l))

    Returns:
        float -- transfer entropy from X to Y
    """
    T = 0
    for (YtXt_list, val) in pY_t_tmk_X_tm1_tml.items():
        Yt_list = YtXt_list[:k]
        logval1 = (val * pY_tm1_tmk[Yt_list[1:]])
        logval2 = (pY_tm1_tmk_X_tm1_tml[YtXt_list[1:]]*pY_t_tmk[Yt_list])
        if logval1 == 0 or logval2 == 0:
            continue
        logval = logval1 / logval2
        T += val * np.log(logval)/np.log(2)

    return T


def TE(Sy, Sx, bins, T, T_func):
    """Computes transfer entropy from X to Y (TE_XY)
    
    Arguments:
        Sy {list of float} -- timestamps of user Y
        Sx {list of float} -- timestamps of user X
        bins {list of int} -- bin widths in seconds, bin[0] = bin width at time t0, bin[1] = bin width at time t-1, bin[2] = bin width at time t-2, ...
        T {int} -- time boundary in hours
        T_func {function} -- which function to use for TE computation
    
    Returns:
        float -- transfer entropy from X to Y
        float -- bias estimation
    """
    Sx.sort()
    Sy.sort()
    k = len(bins)
    prob_map, total = binning.compute_binning(Sy, Sx, bins, T)
    binsX_indices = list(range(k, 2*k))
    pY_t_tmk_X_tm1_tml = binning.eliminate_bin(prob_map, [binsX_indices[0]])
    
    pY_tm1_tmk_X_tm1_tml = binning.eliminate_bin(prob_map, [0]+[binsX_indices[0]])
    pY_t_tmk = binning.eliminate_bin(prob_map, binsX_indices)
    pY_tm1_tmk = binning.eliminate_bin(prob_map, [0]+binsX_indices)
    pt_bias = panzeri_treves_bias(prob_map, total, len(bins), len(bins))

    te = T_func(pY_t_tmk, pY_tm1_tmk, pY_tm1_tmk_X_tm1_tml, pY_t_tmk_X_tm1_tml, k)
    return te, pt_bias
    

File: cpp/src/transfer_entropy.cpp
#include "transfer_entropy.h"
#include "binning.h"
#include <algorithm>
#include <iostream>
#include <numeric>
#include <cstring>
#include <math.h>
#include <bitset>
#include <chrono>
#include <assert.h>
#include <iterator>
#include <iomanip>
using namespace std;



TransferEntropy::TransferEntropy(double* Sy, int len_Sy, double* Sx, int len_Sx) : Sy(Sy), len_Sy(len_Sy), Sx(Sx), len_Sx(len_Sx) {
  
  
}

double TransferEntropy::panzeri_treves_bias(double* pY_t_tmk_X_tm1_tml, double total, int k, double* pY_t_tmk)
{
  int len_pY_t_tmk_X_tm1_tml =  (1 << (k + k - 1));
  int len_pY_t_tmk = (1 << k);
  double N = 0;
  for(int i = 0; i < len_pY_t_tmk_X_tm1_tml; i++)
    N += pY_t_tmk_X_tm1_tml[i] * total;

  double Nb_sum1 = 0;
  for(int i = 0; i < len_pY_t_tmk; i++) 
    if(pY_t_tmk[i]) 
      Nb_sum1 += ((i & (1 << (k-1))) >> (k-1));
  

  double Nb_sum2 = 0;
  for(int i = 0; i < len_pY_t_tmk_X_tm1_tml; i++)
    if(pY_t_tmk_X_tm1_tml[i])
      Nb_sum2 += ((i & (1 << (k+k-2))) >> (k+k-2));


  return (-1 / (2 * N * log(2))) * (Nb_sum1 - Nb_sum2);    
}

void print_prob(double* p, int len)
{
  for(int i = 0; i < len; i++)
    cout << "  " << p[i] << "  " << endl;
  cout << endl;
}

double TransferEntropy::T1(double* pY_t_tmk, double* pY_tm1_tmk, double* pY_tm1_tmk_X_tm1_tml, double* pY_t_tmk_X_tm1_tml, int k)
{

  int len_pY_t_tmk = (1 << k);
  int len_pY_t_tmk_X_tm1_tml = (1 << (k + k - 1));

  double H1 = 0.0;

  for(int i = 0; i < len_pY_t_tmk; i++)
  {
    double val = pY_t_tmk[i];
    double pY_tm1_tmk_val = pY_tm1_tmk[i & ((1 << (k-1)) - 1)];
   
    if(val == 0 || pY_tm1_tmk_val == 0)
      continue;

    
    
    double pcond = val / pY_tm1_tmk_val;
    H1 += val * log2(pcond);
  }
  H1 = -H1;
  HY = H1;

  double H2 = 0;

   for(int i = 0; i < len_pY_t_tmk_X_tm1_tml; i++)
  {
    double val = pY_t_tmk_X_tm1_tml[i];
    double pY_tm1_tmk_X_tm1_tml_val = pY_tm1_tmk_X_tm1_tml[i & ((1 << (k+k-2)) - 1)];
    if(val == 0 || pY_tm1_tmk_X_tm1_tml_val == 0)
      continue;
    


    double pcond = val / pY_tm1_tmk_X_tm1_tml_val;
    H2 += val * log2(pcond);

  }

  H2 = -H2;
  HYX = H2;
  return max(H1 - H2, 0.0);  
}

double TransferEntropy::T2(double* pY_t_tmk, double* pY_tm1_tmk, double* pY_tm1_tmk_X_tm1_tml, double* pY_t_tmk_X_tm1_tml, int k)
{
  double te = 0;
  int len_pY_t_tmk_X_tm1_tml = (1 << (k + k - 1));
  int len_pY_t_tmk = 1 << (k-1);


  for(int i = 0; i < len_pY_t_tmk_X_tm1_tml; i++)
  {
    double val = pY_t_tmk_X_tm1_tml[i]; //ok
    double pY_tm1_tmk_val = pY_tm1_tmk[(i >> (k-1)) & ((1 << (k-1))-1)];
    double pY_tm1_tmk_X_tm1_tml_val = pY_tm1_tmk_X_tm1_tml[i & (~(1 << (k+k-2)))]; //ok
    double pY_t_tmk_val = pY_t_tmk[(i / len_pY_t_tmk)]; //ok

    double logval1 = val * pY_tm1_tmk_val;
    double logval2 = pY_tm1_tmk_X_tm1_tml_val * pY_t_tmk_val;
    if(logval1 == 0 || logval2 == 0)
      continue;
    te += val * log2(logval1 / logval2);
  }


  return max(0.0, te);

}


double TransferEntropy::TE(int* bins, int k, int T, double (TransferEntropy::*T_func)(double*, double*, double*, double*, int), double* pt_bias)
{
    Binning binning;
    vector<double> Sy_filtered;
    vector<double> Sx_filtered;
    for(int i = 0; i < len_Sx; i++)
      if(Sx[i] <= T)
        Sx_filtered.emplace_back(Sx[i]);
    for(int i = 0; i < len_Sy; i++)
      if(Sy[i] <= T)
        Sy_filtered.emplace_back(Sy[i]);
      
    int total = binning.compute_binning(Sy_filtered.data(), Sy_filtered.size(), Sx_filtered.data(), Sx_filtered.size(), bins, k, T);

    int eliminate_bins = (~((1 << k) - 1));
    double* pY_t_tmk = new double[(1 << k)];
    memset(pY_t_tmk, 0, ((1 << k)) * sizeof(double));
    binning.eliminate_bin(eliminate_bins, pY_t_tmk);

    
    
    eliminate_bins = (~((1 << k) - 1)) & (~ (1 << (k+k)-1));
    double* pY_tm1_tmk = new double[1 << (k-1)];
    memset(pY_tm1_tmk, 0, (1 << (k-1)) * sizeof(double));
    binning.eliminate_bin(eliminate_bins, pY_tm1_tmk);
    
    eliminate_bins = (~((1 << k) - 1)) & (~ (1 << (k+k)-1)) | ((1<<(k-1))-1);
    double* pY_tm1_tmk_X_tm1_tml = new double[(1 << (k + k - 2))];
    memset(pY_tm1_tmk_X_tm1_tml, 0, (1 << (k + k - 2)) * sizeof(double));
    binning.eliminate_bin(eliminate_bins, pY_tm1_tmk_X_tm1_tml);



    eliminate_bins = (~((1 << k) - 1)) | ((1<<(k-1))-1);
    double* pY_t_tmk_X_tm1_tml = new double[(1 << (k + k - 1))];
    memset(pY_t_tmk_X_tm1_tml, 0,(1 << (k + k - 1)) * sizeof(double));
    binning.eliminate_bin(eliminate_bins, pY_t_tmk_X_tm1_tml);
  

    double te = (this->*T_func)(pY_t_tmk, pY_tm1_tmk, pY_tm1_tmk_X_tm1_tml, pY_t_tmk_X_tm1_tml, k);

    if(pt_bias)
      *pt_bias = panzeri_treves_bias(pY_t_tmk_X_tm1_tml, total, k, pY_t_tmk);

    delete[] pY_t_tmk;
    delete[] pY_tm1_tmk;
    delete[] pY_tm1_tmk_X_tm1_tml;
    delete[] pY_t_tmk_X_tm1_tml;

    return te;
}